sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/DynamicJoinSelection.scala

```diff
@@ -0,0 +1,81 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.execution.adaptive
+
+import org.apache.spark.sql.catalyst.plans.logical.{HintInfo, Join, JoinStrategyHint, JoinStrategyHintCollection, LogicalPlan, NO_BROADCAST_HASH, PREFER_SHUFFLE_HASH}
+import org.apache.spark.sql.catalyst.rules.Rule
+
+/**
+ * This optimization rule incluede two join selection:
+ *   1. detects a join child that has a high ratio of empty partitions and adds a
+ *      no_broadcast_hash_join hint to avoid it being broadcast.
+ *   2. detects a join child that every partition size less than local map threshold and adds a
+ *      prefer_shuffled_hash hint to encourage being shuffle hash join instead of sort merge join.
+ */
+object DynamicJoinSelection extends Rule[LogicalPlan] {
+
+  private def shouldDemoteBroadcastHashJoin(stage: ShuffleQueryStageExec): Boolean = {
+    val mapStats = stage.mapStats.get
+    val partitionCnt = mapStats.bytesByPartitionId.length
+    val nonZeroCnt = mapStats.bytesByPartitionId.count(_ > 0)
+    partitionCnt > 0 && nonZeroCnt > 0 &&
+      (nonZeroCnt * 1.0 / partitionCnt) < conf.nonEmptyPartitionRatioForBroadcastJoin
+  }
+
+  private def preferShuffledHashJoin(stage: ShuffleQueryStageExec): Boolean = {
+    val localMapThreshold = conf.shuffleHashJoinLocalMapThreshold
+    stage.mapStats.get.bytesByPartitionId.forall(_ <= localMapThreshold)
+  }
+
+  private def selectJoinStrategy(plan: LogicalPlan): Option[JoinStrategyHint] = plan match {
+    case LogicalQueryStage(_, stage: ShuffleQueryStageExec) if stage.resultOption.get().isDefined
+      && stage.mapStats.isDefined =>
+      var strategies = Seq[JoinStrategyHint]()
+      if (shouldDemoteBroadcastHashJoin(stage)) {
+        strategies = strategies ++ Seq(NO_BROADCAST_HASH)
+      }
+      if (preferShuffledHashJoin(stage)) {
```

Comment:
are these two join hints exclusive? shall we use `else if` here?

Reply:
I think `PREFER_SHUFFLE_HASH` means we shouldn't do broadcast join.

Reply:
Not exclusive. My thought is `PREFER_SHUFFLE_HASH` is not a determining hint which is same as `NO_BROADCAST_HASH`. `NO_BROADCAST_HASH` is for `BHJ` and other shuffle based join, and `PREFER_SHUFFLE_HASH` is for `SMJ` and `SHJ`.

So one join can be a `BHJ`  even if we have a `PREFER_SHUFFLE_HASH` hint but never be a `SMJ`. In this idea, we can inject some join recommendation for `JoinSelection`.

Reply:
I think the major difference between this new `PREFER_SHUFFLE_HASH` and `SHUFFLE_HASH` hints is: `PREFER_SHUFFLE_HASH` forces to do shuffle hash join even if we can broadcast.

I'd really like to avoid `JoinStrategyHintCollection`, which adds complexity to the join hint framework.

Reply:
@cloud-fan How about this updating ? Has simplified code about hints.